Sources/PT.PM.AntlrUtils/LightInputStream.cs

```diff
@@ -0,0 +1,158 @@
+using System;
+using System.Runtime.CompilerServices;
+using Antlr4.Runtime;
+using Antlr4.Runtime.Misc;
+using PT.PM.Common.Files;
+
+namespace PT.PM.AntlrUtils
+{
+    public class LightInputStream : ICharStream
+    {
+        private readonly char[] data;
+
+        /// <summary>How many characters are actually in the buffer</summary>
+        private readonly int size;
+
+        /// <summary>0..n-1 index into string of next char</summary>
+        private int position;
+
+        public IVocabulary Vocabulary { get; }
+
+        public CaseInsensitiveType CaseInsensitiveType { get; }
+
+        public TextFile TextFile { get; }
+
+        public LightInputStream(IVocabulary vocabulary, TextFile textFile, string input, CaseInsensitiveType caseInsensitiveType = CaseInsensitiveType.None)
+        {
+            Vocabulary = vocabulary ?? throw new ArgumentNullException(nameof(vocabulary));
+            TextFile = textFile ?? throw new ArgumentNullException(nameof(textFile));
+
+            CaseInsensitiveType = caseInsensitiveType;
+
+            data = new char[input.Length];
+            int i = 0;
+            while (i < input.Length)
+            {
+                if (input[i] == '\r')
+                {
+                    if (i + 1 >= input.Length || input[i + 1] != '\n')
```

Comment:
Can't this condition be merged into previous one like this `(input[i] == '\r') && (i + 1 < input.Length) && (input[i + 1] != '\n')`? Then you can get rid of duplicate else branch. And arguably the intention becomes clearer.

Reply:
LOL, I still got the logic wrong. :smile: 
OK, the translation should be literal: `(input[i] == '\r') && ((i + 1 >= input.Length) || (input[i + 1] != '\n'))`